### PR TITLE
feat: bulk connections export tools

### PR DESCRIPTION
**Pull request by Desperado**
## Summary

Adds two MCP tools for bulk LinkedIn connections export. Rebased onto current `main`.

### `get_my_connections`
Collects the authenticated user's connections via infinite scroll. The list lives in a scrollable inner `<main>` container, so `scroll_connections_container` scrolls that ancestor (not `document.body`), tolerates lazy-load latency, and stops early once `limit` unique profiles are loaded. Returns `{username, name, headline}` per connection.

### `extract_contact_details`
Enriches profiles with structured contact data from the profile page and the contact-info overlay (raw innerText kept as fallback):

| Field | Source |
|-------|--------|
| `first_name`, `last_name` | Profile — the name line above the "Nth degree connection" marker |
| `headline` | Profile — first substantive line after the marker |
| `company` | Profile — line after the headline |
| `location` | Profile — the line carrying the "Contact info" suffix |
| `email`, `phone`, `website`, `birthday` | Contact-info overlay — `Label` line then its value (email also has a regex fallback) |
| `profile_raw`, `contact_info_raw` | Original innerText kept as fallback |

**Rate-limit handling:** profiles are processed in chunked batches with configurable `chunk_size` (default 5) and `chunk_delay` (default 30s). The batch stops early on a hard rate limit and returns partial results with a `rate_limited` flag; individual loads retry once after a 5s backoff on soft rate limits (empty-content responses).

### Files changed (5)
- `linkedin_mcp_server/tools/connections.py` — new tool module (uses `tool_timeout` + `get_ready_extractor`, matching the other tools)
- `linkedin_mcp_server/scraping/extractor.py` — `scrape_connections_list()`, `scrape_contact_batch()`, `_parse_contact_record()`
- `linkedin_mcp_server/core/utils.py` — `scroll_connections_container()` (inner-container infinite scroll)
- `linkedin_mcp_server/server.py` — registers `register_connections_tools(mcp, tool_timeout=...)`
- `tests/test_connections_batch.py` — chunking + parser tests

### Scope note
The earlier revision of this branch also carried a `network` degree filter for `search_people`; that's been **dropped** because `main` now ships its own `network: list[str]` filter. This branch is now connections-only.

## Test plan

- [x] `ruff format`, `ruff check`, `ty check` all clean
- [x] Full suite green: **555 passed** (adds `tests/test_connections_batch.py`)
- [x] Server builds and registers both tools; `get_my_connections` returns connections live (verified `limit=60` → 60 unique, `limit=8` → 8 after rebase)
- [x] `extract_contact_details` returns correct structured fields on live profiles (name, company, location; emails where the member shared one — 2/3)
- [x] Chunking + inter-chunk delays verified deterministically (chunk boundaries, no delay after the final chunk, progress callbacks, hard-stop, soft-skip)
- [ ] Live rate-limit *avoidance* at a 50+ real-profile batch — intentionally not stress-tested against a real account

---

Generated with Claude Opus 4.8.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<h3>Greptile Summary</h3>

This PR adds two new MCP tools for bulk LinkedIn connections export: `get_my_connections` (infinite scroll collection of connection usernames) and `extract_contact_details` (batch enrichment with structured contact data). The implementation follows existing code patterns and includes thoughtful rate-limit handling with configurable chunking and delays.

**Key changes:**
- New connections.py module with two tools following the existing tool registration pattern
- Bulk export methods in extractor.py with chunked batch processing and soft/hard rate-limit handling
- Structured contact parser `_parse_contact_record` that extracts fields like email, phone, location from raw profile text
- ERR_ABORTED navigation error handling for timing edge cases
- Network degree filter (1st/2nd/3rd+) added to `search_people`

**Code quality:** The implementation is well-structured with proper error handling, progress reporting, deduplication, and defensive programming. Most issues from previous review rounds have been addressed.

**Minor issue:** Inconsistent ERR_ABORTED handling between initial navigation and re-navigation (line 559) could cause crashes in rare edge cases where the page navigates away during scroll and re-navigation encounters the same timing issue.

<h3>Confidence Score: 4/5</h3>

- This PR is safe to merge with low risk - well-tested implementation following established patterns
- Score reflects thorough testing (105 tests passing), clean code following existing patterns, and comprehensive rate-limit handling. Deducting 1 point for the ERR_ABORTED inconsistency that could cause failures in rare edge cases where page navigation timing issues occur during re-navigation.
- Pay attention to `linkedin_mcp_server/scraping/extractor.py` - verify the ERR_ABORTED handling inconsistency at line 559 won't impact production usage

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| linkedin_mcp_server/tools/connections.py | New file adding two MCP tools for bulk connection export - `get_my_connections` for scraping connection list and `extract_contact_details` for enriching profiles with contact data. Follows existing tool patterns, includes proper error handling and progress reporting. |
| linkedin_mcp_server/scraping/extractor.py | Adds bulk export methods `scrape_connections_list` and `scrape_contact_batch` with chunked rate-limit handling, plus `_parse_contact_record` parser for structured field extraction. Includes ERR_ABORTED navigation handling and soft rate-limit sentinels. Small `network` filter added to `search_people`. |
| linkedin_mcp_server/server.py | Simple registration of new connections tools module - follows existing pattern for tool registration with no changes to other logic. |
| linkedin_mcp_server/tools/person.py | Adds optional `network` parameter to `search_people` for filtering by connection degree (1st, 2nd, 3rd+) - minimal, focused change with proper parameter passthrough. |

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant Client
    participant Tool as connections.py<br/>(extract_contact_details)
    participant Extractor as extractor.py<br/>(scrape_contact_batch)
    participant Browser as LinkedIn Pages
    
    Client->>Tool: extract_contact_details(usernames, chunk_size, chunk_delay)
    Tool->>Tool: Parse & deduplicate usernames
    Tool->>Extractor: scrape_contact_batch(usernames, chunk_size, chunk_delay)
    
    loop For each chunk
        loop For each username in chunk
            Extractor->>Browser: Navigate to profile page
            Browser-->>Extractor: profile_text
            
            alt Soft rate limit (empty content)
                Extractor->>Extractor: Check for _RATE_LIMITED_MSG sentinel
                Extractor->>Extractor: Skip username, add to failed[]
            else Success
                Extractor->>Browser: Navigate to contact info overlay
                Browser-->>Extractor: contact_text
                Extractor->>Extractor: _parse_contact_record(profile, contact)
                Extractor->>Extractor: Add to contacts[]
            end
            
            alt Hard rate limit (RateLimitError)
                Extractor->>Extractor: Add to failed[], set rate_limited=true
                Extractor->>Extractor: Break loop
            end
        end
        
        Extractor->>Tool: Report progress
        
        alt Not last chunk
            Extractor->>Extractor: Sleep(chunk_delay)
        end
    end
    
    Extractor-->>Tool: {contacts[], total, failed[], rate_limited, pages_visited[]}
    Tool-->>Client: Return enriched data
```

<sub>Last reviewed commit: b9add13</sub>

### CODE REVIEW

File: linkedin_mcp_server/scraping/extractor.py [hunk 1]
```diff
@@ -411,3 +492,192 @@ async def search_people(
...
+
+                    results.push({ username, name, headline });
+                }
+                return results;
+            }"""
+        )
+
+        # Apply limit
+        if limit > 0:
+            raw_connections = raw_connections[:limit]
```
[the remark this answers is not in the document]
Comment (Desperado) [reply linkedin_mcp_server/scraping/extractor.py]: Won't fix — the suggestion to break early in the JS loop wouldn't help because the expensive part is `scroll_to_bottom()`, which runs *before* the JS extraction. By the time the DOM query executes, all scrolling is already done. Users already control scroll depth via the `max_scrolls` parameter (e.g. `max_scrolls=3` for quick results). A proper fix would require refactoring the generic `scroll_to_bottom` utility to accept an early-exit predicate, which is out of scope for this PR.

File: linkedin_mcp_server/tools/connections.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+            Each contact record contains:
+            - username, first_name, last_name: Identity fields
+            - email, phone, website, birthday: From the contact info overlay
+            - headline, location, company: From the main profile page
+            - profile_raw, contact_info_raw: Original innerText as fallback
+        """
+        try:
+            await ensure_authenticated()
+
+            username_list = [u.strip() for u in usernames.split(",") if u.strip()]
```
[the remark this answers is not in the document]
Comment (Desperado) [reply linkedin_mcp_server/tools/connections.py]: Fixed in 35a7643 — added `dict.fromkeys()` deduplication.

File: linkedin_mcp_server/tools/connections.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+            title="Get My Connections",
+            readOnlyHint=True,
+            destructiveHint=False,
+            openWorldHint=True,
+        )
+    )
+    async def get_my_connections(
+        ctx: Context,
+        limit: int = 0,
+        max_scrolls: int = 50,
```
[the remark this answers is not in the document]
Comment (Desperado) [reply linkedin_mcp_server/tools/connections.py:38]: Won't fix — `max_scrolls=-10` simply makes `range(max_scrolls)` produce an empty sequence, meaning no scrolling occurs. This doesn't crash or corrupt data; it just returns whatever connections are visible without scrolling. Since this is an MCP tool parameter with a sensible default of 50, a negative value is an obvious caller mistake that produces a self-explanatory empty result. Adding validation here would be pure noise.

File: linkedin_mcp_server/scraping/extractor.py [hunk 4]
```diff
@@ -411,3 +492,192 @@ async def search_people(
...
+        self,
+        usernames: list[str],
+        chunk_size: int = 5,
+        chunk_delay: float = 30.0,
+        progress_cb: Callable[[int, int], Awaitable[None]] | None = None,
+    ) -> dict[str, Any]:
+        """Enrich a list of profiles with contact details in chunked batches.
+
+        For each username: scrapes main profile + contact_info overlay.
+
```
[the remark this answers is not in the document]
Comment (Desperado) [reply linkedin_mcp_server/scraping/extractor.py:4189]: Fixed in beebf5e — added `failed.append(username)` before the `break`.

File: linkedin_mcp_server/scraping/extractor.py [hunk 5]
```diff
@@ -411,3 +492,192 @@ async def search_people(
...
+        Returns:
+            {connections: [{username, name, headline}, ...], total, url, pages_visited}
+        """
+        url = "https://www.linkedin.com/mynetwork/invite-connect/connections/"
+
+        await self._page.goto(url, wait_until="domcontentloaded", timeout=30000)
+        await detect_rate_limit(self._page)
+
+        try:
```
[the remark this answers is not in the document]
Comment (Desperado) [reply linkedin_mcp_server/scraping/extractor.py:4097]: Fixed in beebf5e — added `if chunk_size <= 0: raise ValueError(...)` guard at the top of `scrape_contact_batch`.

File: linkedin_mcp_server/tools/connections.py [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+            readOnlyHint=True,
+            destructiveHint=False,
+            openWorldHint=True,
+        )
+    )
+    async def extract_contact_details(
+        usernames: str,
+        ctx: Context,
+        chunk_size: int = 5,
+        chunk_delay: int = 30,
```
[the remark this answers is not in the document]
Comment (Desperado) [reply linkedin_mcp_server/tools/connections.py]: Fixed in 354fc80 — `chunk_delay` parameter type changed to `float`.

File: linkedin_mcp_server/tools/connections.py [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+                )
+
+            result = await extractor.scrape_contact_batch(
+                usernames=username_list,
+                chunk_size=chunk_size,
+                chunk_delay=float(chunk_delay),
+                progress_cb=on_progress,
+            )
+
+            await ctx.report_progress(progress=total, total=total, message="Complete")
```
[the remark this answers is not in the document]
Comment (Desperado) [reply linkedin_mcp_server/tools/connections.py]: Fixed in 354fc80 — now uses `result["total"]` instead of reporting 100%.

File: linkedin_mcp_server/scraping/extractor.py [hunk 8]
```diff
@@ -411,3 +411,188 @@ async def search_people(
...
+
+        Args:
+            usernames: List of LinkedIn usernames to enrich.
+            chunk_size: Profiles per chunk before a long pause.
+            chunk_delay: Seconds to pause between chunks.
+            progress_cb: Optional async callback(completed, total) for progress.
+
+        Returns:
+            {contacts: [{username, name, headline, email, phone, location, ...}],
+             total, failed, rate_limited, pages_visited}
```
[the remark this answers is not in the document]
Comment (Desperado) [reply linkedin_mcp_server/scraping/extractor.py]: Fixed in 8194967 — docstring now lists the actual structured fields (`first_name`, `last_name`, `email`, etc.).

File: linkedin_mcp_server/tools/connections.py [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+            username_list = [u.strip() for u in usernames.split(",") if u.strip()]
+
+            if not username_list:
+                return {
+                    "error": "invalid_input",
+                    "message": "No valid usernames provided. Pass comma-separated usernames.",
+                }
+
+            logger.info(
+                "Enriching %d profiles (chunk_size=%d, chunk_delay=%ds)",
```
[the remark this answers is not in the document]
Comment (Desperado) [reply linkedin_mcp_server/tools/connections.py]: Fixed in 8194967 — changed to `%.0fs`.

File: linkedin_mcp_server/tools/connections.py [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+                usernames=username_list,
+                chunk_size=chunk_size,
+                chunk_delay=chunk_delay,
+                progress_cb=on_progress,
+            )
+
+            completed = result["total"]
+            await ctx.report_progress(
+                progress=completed, total=total, message="Complete"
+            )
```
[the remark this answers is not in the document]
Comment (Desperado) [reply linkedin_mcp_server/tools/connections.py:175]: Fixed in 3ab454c — progress message now shows "Stopped early due to rate limit (N/M processed)" when rate-limited.

File: linkedin_mcp_server/scraping/extractor.py [hunk 11]
```diff
@@ -411,3 +492,192 @@ async def search_people(
...
+                    let headline = '';
+                    if (card) {
+                        const headlineEl = card.querySelector(
+                            '.mn-connection-card__occupation, .entity-result__primary-subtitle, span.t-normal'
+                        );
+                        if (headlineEl) headline = headlineEl.innerText.trim();
+                    }
+                    if (!headline && card) {
+                        // Fallback: split card text by newlines, second non-empty line is usually headline
+                        const lines = card.innerText.split('\\n').map(l => l.trim()).filter(Boolean);
```
[the remark this answers is not in the document]
Comment (Desperado) [reply linkedin_mcp_server/scraping/extractor.py]: Fixed in beebf5e — added guards for both `_RATE_LIMITED_MSG` sentinels. Profile sentinel skips the username (added to `failed`), contact overlay sentinel falls back to empty string.

File: linkedin_mcp_server/scraping/extractor.py [hunk 12]
```diff
@@ -61,6 +61,87 @@ def strip_linkedin_noise(text: str) -> str:
...
+            full_name = non_empty[0]
+            parts = full_name.split(None, 1)
+            result["first_name"] = parts[0] if parts else full_name
+            result["last_name"] = parts[1] if len(parts) > 1 else None
+
+        # Find connection degree marker (· 1st, · 2nd, · 3rd)
+        degree_idx: int | None = None
+        for i, ln in enumerate(non_empty):
+            if re.match(r"^·\s*\d+(st|nd|rd)$", ln):
+                degree_idx = i
```
[the remark this answers is not in the document]
Comment (Desperado) [reply linkedin_mcp_server/scraping/extractor.py]: Fixed in beebf5e — regex extended to `r"^·\s*\d+(st|nd|rd|th)\+?$"`.

File: linkedin_mcp_server/tools/connections.py [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+        annotations=ToolAnnotations(
+            title="Get My Connections",
+            readOnlyHint=True,
+            destructiveHint=False,
+            openWorldHint=True,
+        )
+    )
+    async def get_my_connections(
+        ctx: Context,
+        limit: int = 0,
```
[the remark this answers is not in the document]
Comment (Desperado) [reply linkedin_mcp_server/tools/connections.py:37]: Won't fix — `limit=-10` behaving as unlimited is fine. The parameter semantics are "0 = unlimited" and any non-positive value logically means "no limit". This is consistent and not a bug.

File: linkedin_mcp_server/tools/connections.py [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+            readOnlyHint=True,
+            destructiveHint=False,
+            openWorldHint=True,
+        )
+    )
+    async def extract_contact_details(
+        usernames: str,
+        ctx: Context,
+        chunk_size: int = 5,
+        chunk_delay: float = 30.0,
```
[the remark this answers is not in the document]
Comment (Desperado) [reply linkedin_mcp_server/tools/connections.py:97]: Won't fix — `asyncio.sleep()` with a negative value returns immediately (same as 0), which just means "no delay". This is an MCP tool parameter with a sensible default of 30s; passing a negative value is a caller error with a harmless outcome.

File: linkedin_mcp_server/tools/connections.py [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+            title="Extract Contact Details",
+            readOnlyHint=True,
+            destructiveHint=False,
+            openWorldHint=True,
+        )
+    )
+    async def extract_contact_details(
+        usernames: str,
+        ctx: Context,
+        chunk_size: int = 5,
```
[the remark this answers is not in the document]
Comment (Desperado) [reply linkedin_mcp_server/tools/connections.py:96]: Won't fix — the extractor already validates `chunk_size <= 0` with a clear `ValueError`. Duplicating validation at the tool layer adds no value; the error message from the extractor (`"chunk_size must be a positive integer, got 0"`) is already user-friendly.

**Comment by Desperado** [comment]
The Greptile **"too many files changed (119 found / 100 limit)"** status above isn't this feature — it's branch divergence. `feature/bulk-connections-export` is currently **708 commits ahead and 1,008 behind `main`** (status: *diverged*), so GitHub computes a 119-file / +18.5k diff and Greptile skips the review at its 100-file limit.

The actual feature footprint is ~7 files:

- `linkedin_mcp_server/tools/connections.py`
- `linkedin_mcp_server/scraping/extractor.py`
- `linkedin_mcp_server/core/utils.py`
- `linkedin_mcp_server/server.py`
- `linkedin_mcp_server/tools/person.py`
- `tests/test_connections_batch.py`, `tests/test_tools.py`

Rebasing the branch onto current `main` should drop the diff back to those files and let Greptile review (and give human reviewers a clean diff). Happy to do the rebase if that's the preferred path.

**Comment by Desperado** [comment]
Done — rebased onto current `main`. The branch is now **2 commits ahead / 0 behind** (was 708/1008 diverged), the diff is **5 files / +830** (was 119 files), and GitHub reports it **mergeable**. That should put it under Greptile's 100-file limit so the review can run.

Two notes on the rebase:
- The previously-separate `search_people` network-degree filter was **dropped**, since `main` now ships its own `network: list[str]` filter.
- The earlier P0 (`extract_page`/`_extract_overlay` called without `section_name`, `ExtractedSection` treated as a string) is fixed — calls pass `section_name` and read `.text`. Full suite is green (555 passed) and both tools were smoke-tested live.

**Comment by Desperado** [comment]
Thanks — addressed in `1ea9d0d`:

**P1 — layout class names in `scrape_connections_list` (fixed).** Dropped all `.mn-connection-card__*` / `.entity-result__*` / `span.t-bold` / `span.t-normal` queries. Name/headline are now derived structurally from the card's innerText (name = first line, headline = longest remaining line). No layout classes; verified live (names and headlines still correct).

**P1 — locale-dependent anchors in `_parse_contact_record` (mitigated + documented).** Fully locale-independent parsing of name/headline/company/location from innerText isn't really achievable (there are no stable structural anchors once you drop the localized markers). What I've done instead: email is already recovered locale-independently via regex, and `profile_raw`/`contact_info_raw` are always returned as a language-agnostic fallback, so it's graceful degradation rather than a hard failure. I've documented the English-UI assumption on the parser. Happy to add per-locale marker tables in a follow-up if you'd like full coverage.

**P1 — return schema lacks `sections` (documented; deferring the call to you).** These bulk tools emit per-record lists, which don't map cleanly onto `{url, sections: {name: raw_text}}` (that shape is for page-section raw text). I've documented the intentional deviation in the tool docstrings. If you'd prefer them to conform — e.g. `sections: {"connections": <json>}` with `connections`/`contacts` as a convenience alias — say the word and I'll wrap them; I didn't want to force an awkward shape without your sign-off.

**Comment by Desperado** [comment]
Good catch — fixed in `829a48a`. Added `await detect_rate_limit(self._page)` right after the re-navigation `goto`, mirroring the initial-navigation check, so a challenge/checkpoint reached during re-navigation now raises `RateLimitError` instead of silently returning zero connections.

**Comment by Desperado** [comment]
@stickerdaniel please check this on your availablity

**Comment by Desperado** [comment]
Note: closed the duplicate draft PR #486 (`feature/bulk-connections-export-upstream`) as superseded by this one. #170 is the single, current version of the bulk-connections feature.
